File: sliding-window/LongestSubstring.java
import java.util.ArrayList;

public class LongestSubstring {
    public static int getLongestSubstring(String str){
        int maxLen = 0;
        int r =0;
        int l = 0;
        int n = str.length();
        ArrayList<Integer> vis = new ArrayList<>();
        for(int i =0;i<256;i++){
            vis.add(-1);
        }
        while (r<n) {
            if(vis.get(str.charAt(r))!= -1 && vis.get(str.charAt(r)) >= l){
                l = vis.get(str.charAt(r)) + 1;
            }
            maxLen = Integer.max(maxLen, r - l + 1);
            vis.set(str.charAt(r), r);
            r++;
        }
        return maxLen;
    }
    public static void main(String[] args) {
        String str = "abcabcbb";
        System.out.println(getLongestSubstring(str));
    }
}

File: sliding-window/longest_substring.cpp
#include<iostream>
#include<string>
#include<vector>

int get_longest_substring(std::string &str){
    int left_ptr = 0;
    int right_ptr = 0;
    int max_len = 0;
    int n = str.length();
    std::vector<int>vis(256,-1);
    while(right_ptr < n){
        if(vis[str[right_ptr]] != -1 && vis[str[right_ptr]] >=left_ptr){
            left_ptr = vis[str[right_ptr]] + 1;
        }
        max_len = std::max(max_len , right_ptr - left_ptr + 1 );
        vis[str[right_ptr]] = right_ptr;
        right_ptr++;
    }
    return max_len;
}

int main(){
    std::string str = "abcabcbb";
    std::cout<<get_longest_substring(str);
    return 0;
}